### PR TITLE
Further room version wiring

### PR DESCRIPTION
**Pull request by neilalexander**
This PR contains some more wiring for room versions.

### CODE REVIEW

File: roomserver/alias/alias.go [hunk 1]
```diff
@@ -250,10 +250,18 @@ func (r *RoomserverAliasAPI) sendUpdatedAliasesEvent(
...
+	if err != nil {
+		return err
+	}
+	if eventIDFormat == gomatrixserverlib.EventIDFormatV1 {
+		eventID = fmt.Sprintf("$%s:%s", util.RandomString(16), r.Cfg.Matrix.ServerName)
+	}
 	now := time.Now()
 	event, err := builder.Build(
-		eventID, now, r.Cfg.Matrix.ServerName, r.Cfg.Matrix.KeyID, r.Cfg.Matrix.PrivateKey,
+		eventID, now, r.Cfg.Matrix.ServerName, r.Cfg.Matrix.KeyID,
```
Comment (Kegsay) [inline roomserver/alias/alias.go]: Yeah, see what I mean about how every caller has to do the same thing? :)

File: syncapi/routing/messages.go [hunk 2]
```diff
@@ -380,17 +386,20 @@ func (r *messagesReq) backfill(fromEventIDs []string, limit int) ([]gomatrixserv
...
 
 	// Store the events in the database, while marking them as unfit to show
 	// up in responses to sync requests.
-	for _, pdu := range pdus {
-		headered := pdu.Headered(gomatrixserverlib.RoomVersionV1)
+	for i := range pdus {
 		if _, err = r.db.WriteEvent(
 			r.ctx,
-			&headered,
+			&pdus[i],
```
Comment (Kegsay) [inline syncapi/routing/messages.go]: I dislike this because it's not clear what form 'PDU' is:
 - An `Event`
 - A `json.RawMessage`
 - A `HeaderedEvent`

Please rename to be clearer.

File: roomserver/storage/sqlite3/state_block_table.go [hunk 3]
```diff
@@ -39,7 +39,8 @@ const stateDataSchema = `
 
 const insertStateDataSQL = "" +
 	"INSERT INTO roomserver_state_block (state_block_nid, event_type_nid, event_state_key_nid, event_nid)" +
-	" VALUES ($1, $2, $3, $4)"
+	" VALUES ($1, $2, $3, $4)" +
+	" ON CONFLICT DO NOTHING"
```
Comment (Kegsay) [inline roomserver/storage/sqlite3/state_block_table.go]: Why did you add this?

File: federationsender/queue/destinationqueue.go [hunk 4]
```diff
@@ -119,7 +123,7 @@ func (oq *destinationQueue) next() *gomatrixserverlib.Transaction {
 	oq.lastTransactionIDs = []gomatrixserverlib.TransactionID{t.TransactionID}
 
 	for _, pdu := range oq.pendingEvents {
-		t.PDUs = append(t.PDUs, *pdu)
+		t.PDUs = append(t.PDUs, (*pdu).JSON())
```
Comment (Kegsay) [inline federationsender/queue/destinationqueue.go]: I'm unclear what form this takes. Is it the raw event JSON?

File: clientapi/routing/createroom.go [hunk 5]
```diff
@@ -378,8 +379,18 @@ func buildEvent(
...
-	event, err := builder.Build(eventID, evTime, cfg.Matrix.ServerName, cfg.Matrix.KeyID, cfg.Matrix.PrivateKey)
+	eventID := ""
+	eventIDFormat, err := roomVersion.EventIDFormat()
+	if err != nil {
+		return nil, err
+	}
+	if eventIDFormat == gomatrixserverlib.EventIDFormatV1 {
+		eventID = fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.Matrix.ServerName)
+	}
+	event, err := builder.Build(
```
Comment (Kegsay) [inline clientapi/routing/createroom.go]: Once again, the same code..

File: federationapi/routing/threepid.go [hunk 6]
```diff
@@ -280,9 +310,17 @@ func buildMembershipEvent(
...
+	eventIDFormat, err := queryRes.RoomVersion.EventIDFormat()
+	if err != nil {
+		return nil, err
+	}
+	if eventIDFormat == gomatrixserverlib.EventIDFormatV1 {
+		eventID = fmt.Sprintf("$%s:%s", util.RandomString(16), cfg.Matrix.ServerName)
+	}
 	event, err := builder.Build(
-		eventID, time.Now(), cfg.Matrix.ServerName, cfg.Matrix.KeyID, cfg.Matrix.PrivateKey,
+		eventID, time.Now(), cfg.Matrix.ServerName, cfg.Matrix.KeyID,
```
Comment (Kegsay) [inline federationapi/routing/threepid.go]: And again..

File: federationapi/routing/join.go [hunk 7]
```diff
@@ -88,9 +99,14 @@ func MakeJoin(
...
 
+	resultMap := map[string]interface{}{
+		"event":        builder,
+		"room_version": verRes.RoomVersion,
+	}
+
 	return util.JSONResponse{
 		Code: http.StatusOK,
-		JSON: map[string]interface{}{"event": builder},
+		JSON: resultMap,
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply federationapi/routing/join.go]: Fixed

File: clientapi/routing/joinroom.go [hunk 8]
```diff
@@ -312,17 +330,37 @@ func (r joinRoomReq) joinRoomUsingServer(roomID string, server gomatrixserverlib
...
+	if err != nil {
+		return nil, err
+	}
+	eventID := ""
+	if eventIDFormat == gomatrixserverlib.EventIDFormatV1 {
+		eventID = fmt.Sprintf("$%s:%s", util.RandomString(16), r.cfg.Matrix.ServerName)
+	}
 	event, err := respMakeJoin.JoinEvent.Build(
-		eventID, r.evTime, r.cfg.Matrix.ServerName, r.cfg.Matrix.KeyID, r.cfg.Matrix.PrivateKey,
+		eventID, r.evTime, r.cfg.Matrix.ServerName, r.cfg.Matrix.KeyID,
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply clientapi/routing/joinroom.go]: Done

File: clientapi/routing/joinroom.go [hunk 9]
```diff
@@ -312,17 +330,37 @@ func (r joinRoomReq) joinRoomUsingServer(roomID string, server gomatrixserverlib
 		return nil, err
 	}
 
-	eventID := fmt.Sprintf("$%s:%s", util.RandomString(16), r.cfg.Matrix.ServerName)
+	if respMakeJoin.RoomVersion == "" {
+		respMakeJoin.RoomVersion = gomatrixserverlib.RoomVersionV1
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply clientapi/routing/joinroom.go]: EventBuilder no longer takes event ID so done

File: clientapi/consumers/roomserver.go [hunk 10]
```diff
@@ -71,6 +71,7 @@ func (s *OutputRoomEventConsumer) Start() error {
 func (s *OutputRoomEventConsumer) onMessage(msg *sarama.ConsumerMessage) error {
 	// Parse out the event JSON
 	var output api.OutputEvent
+	output.NewRoomEvent = &api.OutputNewRoomEvent{}
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply clientapi/consumers/roomserver.go]: This was a part of my debugging - it's gone now

File: federationapi/routing/join.go [hunk 11]
```diff
@@ -104,8 +120,17 @@ func SendJoin(
...
 ) util.JSONResponse {
-	var event gomatrixserverlib.Event
-	if err := json.Unmarshal(request.Content(), &event); err != nil {
+	verReq := api.QueryRoomVersionForRoomRequest{RoomID: roomID}
+	verRes := api.QueryRoomVersionForRoomResponse{}
+	if err := query.QueryRoomVersionForRoom(httpReq.Context(), &verReq, &verRes); err != nil {
+		return util.JSONResponse{
+			Code: http.StatusInternalServerError,
+			JSON: jsonerror.InternalServerError(),
+		}
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply federationapi/routing/join.go]: Fixed

File: federationapi/routing/join.go [hunk 12]
```diff
@@ -137,9 +162,10 @@ func SendJoin(
 	}
 
 	// Check that the event is signed by the server sending the request.
+	redacted := event.Redact()
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply federationapi/routing/join.go]: This one needs to be separate to avoid `JSON is not in method set of gomatrixserverlib.Event`. The function takes a pointer receiver but returns a copy.

File: syncapi/routing/messages.go (hunk 2 above)
Comment (neilalexander) [reply syncapi/routing/messages.go]: Named it back to `headered`.

File: roomserver/storage/sqlite3/state_block_table.go (hunk 3 above)
Comment (neilalexander) [reply roomserver/storage/sqlite3/state_block_table.go]: At this point I was trying to debug a problem in my database and should have removed this. Will fix.

File: common/events.go [hunk 13]
```diff
@@ -99,7 +115,32 @@ func AddPrevEventsToEvent(
...
+		}
+		builder.AuthEvents = v2AuthRefs
+		builder.PrevEvents = v2PrevRefs
+	}
 
 	return nil
 }
+
+// RemoveEventHeaders unwraps an array of headered events.
+func RemoveEventHeaders(in []gomatrixserverlib.HeaderedEvent) []gomatrixserverlib.Event {
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply common/events.go]: Done

**Comment by Kegsay** [comment]
```

Test 259 Guest non-joined users can get individual state for world_readable rooms after leaving... 2020-03-27 16:02:31.954 UTC [2458] ERROR:  duplicate key value violates unique constraint "roomserver_state_block_state_block_nid_event_type_nid_event_key"
--


```
